src/main/java/mate/academy/springbootintro/service/shoppingcart/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,130 @@
+package mate.academy.springbootintro.service.shoppingcart.impl;
+
+import jakarta.persistence.EntityNotFoundException;
+import lombok.RequiredArgsConstructor;
+import mate.academy.springbootintro.dto.cartitem.CreateCartItemRequestDto;
+import mate.academy.springbootintro.dto.cartitem.UpdateCartItemRequestDto;
+import mate.academy.springbootintro.dto.shoppingcartdto.ShoppingCardDto;
+import mate.academy.springbootintro.mapper.ShoppingCardMapper;
+import mate.academy.springbootintro.model.Book;
+import mate.academy.springbootintro.model.CartItem;
+import mate.academy.springbootintro.model.ShoppingCart;
+import mate.academy.springbootintro.repository.book.BookRepository;
+import mate.academy.springbootintro.repository.cartitem.CartItemRepository;
+import mate.academy.springbootintro.repository.shoppingcart.ShoppingCartRepository;
+import mate.academy.springbootintro.service.shoppingcart.ShoppingCartService;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final BookRepository bookRepository;
+    private final CartItemRepository cartItemRepository;
+    private final ShoppingCardMapper shoppingCardMapper;
+
+    @Override
+    @Transactional
+    public ShoppingCardDto findShoppingCardByUserId(Long id) {
+        return shoppingCardMapper.toDto(
+                shoppingCartRepository.findByUserId(id)
+                        .orElseThrow(
+                                () -> new EntityNotFoundException(
+                                        "Cannot find shopping cart by user with id: "
+                                                + id
+                                )
+                        )
+        );
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCardDto addCartItem(
+            CreateCartItemRequestDto cartItemRequestDto,
+            Long currentUserId
+    ) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(
+                currentUserId
+        ).orElseThrow(
+                () -> new EntityNotFoundException(
+                        "Cannot found shopping card with user id: "
+                                + currentUserId
+                )
+        );
+        Book book = bookRepository.findById(
+                cartItemRequestDto.bookId()
+        ).orElseThrow(
+                () -> new EntityNotFoundException(
+                        "Cannot found book with id: "
+                                + cartItemRequestDto.bookId()
+                )
+        );
+        CartItem cartItem = cartItemRepository.findByShoppingCartAndBook(
```

Comment:
```suggestion
        shoppingCart.getCartItems().stream()
                .filter(item -> item.getBook().getId().equals(cartItemRequestDto.bookId()))
                .findFirst()
                .ifPresentOrElse(item -> item.setQuantity(item.getQuantity() + cartItemRequestDto.quantity()),
                        () -> addCartItemToCart(cartItemRequestDto, book, cart));
        shoppingCartRepository.save(shoppingCart);
        return shoppingCardMapper.toDto(shoppingCart);
```

Where
```
private void addCartItemToCart(CartItemRequestDto itemDto, Book book, ShoppingCart cart) {
        CartItem cartItem = itemMapper.toCartItem(itemDto);
        cartItem.setBook(book);
        cart.addItemToCart(cartItem);
    }
```

Where in ShoppingCart
```
public void addItemToCart(CartItem cartItem) {
        cartItem.setShoppingCart(this);
        cartItems.add(cartItem);
    }
```

Reply:
not fixed